File: examples/crypto/bls12-381/proof-of-possession.ts
import { randomBytes } from "node:crypto";
import { bls12_381 } from "@noble/curves/bls12-381.js";

// Honest validator
const honestPrivKey = 12345n;
const honestPubKey = bls12_381.G2.Point.BASE.multiply(honestPrivKey);

// Target victim
const victimPrivKey = 99999n;
const victimPubKey = bls12_381.G2.Point.BASE.multiply(victimPrivKey);

// Attacker creates rogue key
const roguePubKey = victimPubKey.subtract(honestPubKey);

// Aggregate keys
const aggregatedPubKey = honestPubKey.add(roguePubKey);

// Without PoP, attacker can forge signatures!
const message = new TextEncoder().encode("Transfer 32 ETH");
const msgHash = bls12_381.G1.hashToCurve(message, {
	DST: "BLS_SIG_BLS12381G1_XMD:SHA-256_SSWU_RO_NUL_",
});

// Only honest validator signs (attacker doesn't sign at all)
const honestSig = msgHash.multiply(honestPrivKey);

// Attacker claims "aggregate signature" is just the honest signature
// Verification will pass because aggPubKey = victimPubKey!
const lhs = bls12_381.pairing(honestSig, bls12_381.G2.Point.BASE);
const rhs = bls12_381.pairing(msgHash, aggregatedPubKey);

function generatePoP(
	privateKey: bigint,
	publicKey: typeof bls12_381.G2.Point.BASE,
): Uint8Array {
	// Sign the public key itself with domain separation
	const pubKeyBytes = publicKey.toRawBytes(false); // Uncompressed
	const popHash = bls12_381.G1.hashToCurve(pubKeyBytes, {
		DST: "BLS_POP_BLS12381G1_XMD:SHA-256_SSWU_RO_POP_",
	});
	const pop = popHash.multiply(privateKey);
	return pop.toRawBytes(true); // Compressed signature
}

const validator1PrivKey =
	BigInt(`0x${randomBytes(32).toString("hex")}`) % bls12_381.fields.Fr.ORDER;
const validator1PubKey = bls12_381.G2.Point.BASE.multiply(validator1PrivKey);
const validator1PoP = generatePoP(validator1PrivKey, validator1PubKey);

function verifyPoP(
	publicKey: typeof bls12_381.G2.Point.BASE,
	popBytes: Uint8Array,
): boolean {
	try {
		// Deserialize PoP signature
		const pop = bls12_381.G1.Point.fromHex(popBytes);

		// Hash public key to G1
		const pubKeyBytes = publicKey.toRawBytes(false);
		const popHash = bls12_381.G1.hashToCurve(pubKeyBytes, {
			DST: "BLS_POP_BLS12381G1_XMD:SHA-256_SSWU_RO_POP_",
		});

		// Verify: e(PoP, G2) = e(H(pubkey), pubkey)
		const lhs = bls12_381.pairing(pop, bls12_381.G2.Point.BASE);
		const rhs = bls12_381.pairing(popHash, publicKey);

		return lhs.equals(rhs);
	} catch {
		return false;
	}
}

const validator1Valid = verifyPoP(validator1PubKey, validator1PoP);

// Try to verify with wrong public key (should fail)
const wrongPubKey = bls12_381.G2.Point.BASE.multiply(54321n);
const wrongValid = verifyPoP(wrongPubKey, validator1PoP);

// Attacker cannot generate valid PoP for rogue key
const roguePopValid = verifyPoP(roguePubKey, validator1PoP);

interface ValidatorDeposit {
	pubkey: Uint8Array; // 48 bytes (compressed G2)
	withdrawalCredentials: Uint8Array; // 32 bytes
	amount: bigint; // 32 ETH in Gwei
	signature: Uint8Array; // 96 bytes (PoP)
}

const GWEI = 1_000_000_000n;
const depositAmount = 32n * GWEI * 1_000_000_000n; // 32 ETH

const deposit: ValidatorDeposit = {
	pubkey: validator1PubKey.toRawBytes(true),
	withdrawalCredentials: randomBytes(32),
	amount: depositAmount,
	signature: validator1PoP,
};

// Deposit contract verifies PoP
const depositPubKey = bls12_381.G2.Point.fromHex(deposit.pubkey);
const depositValid = verifyPoP(depositPubKey, deposit.signature);

const validators = Array.from({ length: 5 }, (_, i) => {
	const privKey = BigInt(10000 + i * 1111);
	const pubKey = bls12_381.G2.Point.BASE.multiply(privKey);
	const pop = generatePoP(privKey, pubKey);
	return { privKey, pubKey, pop };
});
validators.forEach((v, i) => {
	const popValid = verifyPoP(v.pubKey, v.pop);
});

// All validators with verified PoP can safely aggregate
const safeMessage = new TextEncoder().encode("Safe to aggregate");
const safeMsgHash = bls12_381.G1.hashToCurve(safeMessage, {
	DST: "BLS_SIG_BLS12381G1_XMD:SHA-256_SSWU_RO_NUL_",
});

const safeSigs = validators.map((v) => safeMsgHash.multiply(v.privKey));
let aggSafeSig = safeSigs[0];
for (let i = 1; i < safeSigs.length; i++) {
	aggSafeSig = aggSafeSig.add(safeSigs[i]);
}

let aggSafePubKey = validators[0].pubKey;
for (let i = 1; i < validators.length; i++) {
	aggSafePubKey = aggSafePubKey.add(validators[i].pubKey);
}

const safeLhs = bls12_381.pairing(aggSafeSig, bls12_381.G2.Point.BASE);
const safeRhs = bls12_381.pairing(safeMsgHash, aggSafePubKey);
const safeValid = safeLhs.equals(safeRhs);

const popGenerationTime = 0.1; // ~100 μs
const popVerificationTime = 2; // ~2 ms (2 pairings)
const popSize = 48; // bytes

const DST_POP = "BLS_POP_BLS12381G1_XMD:SHA-256_SSWU_RO_POP_";
const DST_SIG = "BLS_SIG_BLS12381G1_XMD:SHA-256_SSWU_RO_NUL_";

File: packages/voltaire-zig/examples/crypto/bls12-381/proof-of-possession.zig
const std = @import("std");
const crypto = @import("crypto");

/// Proof of Possession (PoP)
///
/// Demonstrates BLS proof-of-possession to prevent rogue key attacks:
/// - Understanding rogue key attack vectors
/// - PoP generation and verification
/// - Ethereum validator deposit workflow
/// - Security guarantees
pub fn main() !void {
    const stdout = std.io.getStdOut().writer();

    try stdout.print("=== BLS Proof of Possession (PoP) ===\n\n", .{});

    // 1. The Rogue Key Attack
    try stdout.print("1. The Rogue Key Attack Problem\n", .{});
    try stdout.print("{s}\n", .{"-" ** 60});

    try stdout.print("Scenario:\n", .{});
    try stdout.print("  - Honest validator has pubkey_H = privkey_H * G2\n", .{});
    try stdout.print("  - Victim has pubkey_V = privkey_V * G2\n", .{});
    try stdout.print("  - Attacker chooses: pubkey_A = pubkey_V - pubkey_H\n", .{});

    try stdout.print("\nAggregation:\n", .{});
    try stdout.print("  aggPubKey = pubkey_H + pubkey_A\n", .{});
    try stdout.print("            = pubkey_H + (pubkey_V - pubkey_H)\n", .{});
    try stdout.print("            = pubkey_V\n", .{});

    try stdout.print("\nAttack:\n", .{});
    try stdout.print("  - Honest validator signs message: sig_H = privkey_H * H(msg)\n", .{});
    try stdout.print("  - Attacker submits sig_H as \"aggregate signature\"\n", .{});
    try stdout.print("  - Verification: e(sig_H, G2) = e(H(msg), aggPubKey)\n", .{});
    try stdout.print("  - Since aggPubKey = pubkey_V, signature appears valid!\n", .{});
    try stdout.print("  - Attacker forged signature for victim without knowing privkey_V\n\n", .{});

    // 2. Proof of Possession Solution
    try stdout.print("2. Proof of Possession Solution\n", .{});
    try stdout.print("{s}\n", .{"-" ** 60});

    try stdout.print("PoP Definition:\n", .{});
    try stdout.print("  PoP = Sign(pubkey, privkey)\n", .{});
    try stdout.print("      = privkey * H(pubkey)\n", .{});

    try stdout.print("\nVerification:\n", .{});
    try stdout.print("  e(PoP, G2) = e(H(pubkey), pubkey)\n", .{});

    try stdout.print("\nWhy it works:\n", .{});
    try stdout.print("  - Attacker cannot compute PoP for rogue key\n", .{});
    try stdout.print("  - Would need: PoP_A = privkey_A * H(pubkey_A)\n", .{});
    try stdout.print("  - But privkey_A = privkey_V - privkey_H is unknown!\n", .{});
    try stdout.print("  - Attacker doesn't know privkey_V, so cannot compute PoP_A\n\n", .{});

    // 3. PoP Generation
    try stdout.print("3. PoP Generation Process\n", .{});
    try stdout.print("{s}\n", .{"-" ** 60});

    try stdout.print("Step 1: Generate key pair\n", .{});
    try stdout.print("  privkey = random 32 bytes (scalar in Fr)\n", .{});
    try stdout.print("  pubkey = privkey * G2 (256-byte G2 point)\n", .{});

    try stdout.print("\nStep 2: Serialize public key\n", .{});
    try stdout.print("  pubkey_bytes = serialize(pubkey) // 192 bytes uncompressed\n", .{});

    try stdout.print("\nStep 3: Hash public key to G1\n", .{});
    try stdout.print("  DST = \"BLS_POP_BLS12381G1_XMD:SHA-256_SSWU_RO_POP_\"\n", .{});
    try stdout.print("  pop_hash = hash_to_curve(pubkey_bytes, DST)\n", .{});

    try stdout.print("\nStep 4: Sign with private key\n", .{});
    try stdout.print("  PoP = privkey * pop_hash (G1 scalar multiplication)\n", .{});
    try stdout.print("  PoP_bytes = serialize(PoP) // 48 bytes compressed\n\n", .{});

    // 4. PoP Verification
    try stdout.print("4. PoP Verification Process\n", .{});
    try stdout.print("{s}\n", .{"-" ** 60});

    try stdout.print("Step 1: Deserialize PoP and pubkey\n", .{});
    try stdout.print("  PoP = deserialize_g1(PoP_bytes)\n", .{});
    try stdout.print("  pubkey = deserialize_g2(pubkey_bytes)\n", .{});
    try stdout.print("  Validate: points on curve and in subgroup\n", .{});

    try stdout.print("\nStep 2: Hash public key to G1\n", .{});
    try stdout.print("  DST = \"BLS_POP_BLS12381G1_XMD:SHA-256_SSWU_RO_POP_\"\n", .{});
    try stdout.print("  pop_hash = hash_to_curve(pubkey_bytes, DST)\n", .{});

    try stdout.print("\nStep 3: Pairing check\n", .{});
    try stdout.print("  e(PoP, G2_generator) = e(pop_hash, pubkey)\n", .{});
    try stdout.print("  Rearranged: e(PoP, G2) * e(-pop_hash, pubkey) = 1\n", .{});

    try stdout.print("\nStep 4: Accept or reject\n", .{});
    try stdout.print("  If check passes: Validator proven to know privkey\n", .{});
    try stdout.print("  If check fails: Reject validator deposit\n\n", .{});

    // 5. Ethereum Deposit Contract
    try stdout.print("5. Ethereum Validator Deposit\n", .{});
    try stdout.print("{s}\n", .{"-" ** 60});

    try stdout.print("Deposit Data Structure:\n", .{});
    try stdout.print("  struct DepositData {{\n", .{});
    try stdout.print("    pubkey: [48]u8,                  // Compressed G2 point\n", .{});
    try stdout.print("    withdrawal_credentials: [32]u8, // Withdrawal address\n", .{});
    try stdout.print("    amount: u64,                     // 32 ETH in Gwei\n", .{});
    try stdout.print("    signature: [96]u8,               // PoP (uncompressed)\n", .{});
    try stdout.print("  }}\n", .{});

    try stdout.print("\nDeposit Contract Logic:\n", .{});
    try stdout.print("  1. Receive 32 ETH + deposit data\n", .{});
    try stdout.print("  2. Verify PoP signature against pubkey\n", .{});
    try stdout.print("  3. If invalid: Revert transaction\n", .{});
    try stdout.print("  4. If valid: Emit DepositEvent\n", .{});
    try stdout.print("  5. Beacon chain processes event\n", .{});
    try stdout.print("  6. Validator activated after processing\n\n", .{});

    // 6. Cost Analysis
    try stdout.print("6. PoP Cost Analysis\n", .{});
    try stdout.print("{s}\n", .{"-" ** 60});

    try stdout.print("One-time costs (per validator):\n", .{});
    try stdout.print("  - PoP generation: ~100 μs (offline)\n", .{});
    try stdout.print("  - PoP verification: ~2 ms (at deposit)\n", .{});
    try stdout.print("  - PoP storage: 48 bytes (negligible)\n", .{});

    try stdout.print("\nOngoing benefits:\n", .{});
    try stdout.print("  - Safe signature aggregation (forever)\n", .{});
    try stdout.print("  - No per-signature overhead\n", .{});
    try stdout.print("  - Enables light clients\n", .{});
    try stdout.print("  - Prevents all rogue key attacks\n\n", .{});

    // 7. Domain Separation
    try stdout.print("7. Domain Separation\n", .{});
    try stdout.print("{s}\n", .{"-" ** 60});

    try stdout.print("PoP uses separate DST from regular signatures:\n", .{});
    try stdout.print("\nPoP DST:\n", .{});
    try stdout.print("  BLS_POP_BLS12381G1_XMD:SHA-256_SSWU_RO_POP_\n", .{});

    try stdout.print("\nSignature DSTs (examples):\n", .{});
    try stdout.print("  BLS_SIG_BLS12381G1_XMD:SHA-256_SSWU_RO_POP_BEACON_BLOCK_\n", .{});
    try stdout.print("  BLS_SIG_BLS12381G1_XMD:SHA-256_SSWU_RO_POP_ATTESTATION_\n", .{});

    try stdout.print("\nSeparation prevents:\n", .{});
    try stdout.print("  - Using PoP as regular signature\n", .{});
    try stdout.print("  - Cross-context signature replay\n", .{});
    try stdout.print("  - Protocol confusion attacks\n\n", .{});

    // 8. Security Guarantees
    try stdout.print("8. Security Guarantees\n", .{});
    try stdout.print("{s}\n", .{"-" ** 60});

    try stdout.print("With PoP requirement:\n", .{});
    try stdout.print("  ✓ Rogue key attacks prevented\n", .{});
    try stdout.print("  ✓ Safe to aggregate signatures\n", .{});
    try stdout.print("  ✓ No additional checks needed at aggregation\n", .{});
    try stdout.print("  ✓ Public keys can be freely combined\n", .{});

    try stdout.print("\nWithout PoP (insecure):\n", .{});
    try stdout.print("  ✗ Vulnerable to rogue key attacks\n", .{});
    try stdout.print("  ✗ Cannot safely aggregate\n", .{});
    try stdout.print("  ✗ Alternative: complex multi-round protocols\n", .{});
    try stdout.print("  ✗ Alternative: message-dependent keys (slow)\n\n", .{});

    // 9. Implementation Notes
    try stdout.print("9. Implementation Notes\n", .{});
    try stdout.print("{s}\n", .{"-" ** 60});

    try stdout.print("Using BLS12-381 precompiles:\n", .{});

    try stdout.print("\nPoP Generation:\n", .{});
    try stdout.print("  1. MapFp2ToG2: privkey -> pubkey (G2 mul)\n", .{});
    try stdout.print("  2. Hash pubkey bytes to G1 (hash-to-curve)\n", .{});
    try stdout.print("  3. G1 mul: PoP = privkey * pop_hash\n", .{});

    try stdout.print("\nPoP Verification:\n", .{});
    try stdout.print("  1. Deserialize and validate points\n", .{});
    try stdout.print("  2. Hash pubkey bytes to G1\n", .{});
    try stdout.print("  3. Pairing check (2 pairs, 768 bytes input)\n", .{});

    try stdout.print("\nGas costs (if on-chain):\n", .{});
    try stdout.print("  - Hash-to-curve: ~55,000 gas\n", .{});
    try stdout.print("  - Pairing check: ~161,000 gas (base + 2 pairs)\n", .{});
    try stdout.print("  - Total: ~216,000 gas per PoP verification\n\n", .{});

    // 10. Production Checklist
    try stdout.print("10. Production Checklist\n", .{});
    try stdout.print("{s}\n", .{"-" ** 60});

    try stdout.print("Validator Setup:\n", .{});
    try stdout.print("  [ ] Generate privkey from secure random source (32 bytes)\n", .{});
    try stdout.print("  [ ] Derive pubkey = privkey * G2\n", .{});
    try stdout.print("  [ ] Generate PoP = Sign(pubkey, privkey)\n", .{});
    try stdout.print("  [ ] Verify PoP locally before submitting deposit\n", .{});
    try stdout.print("  [ ] Backup privkey securely (encrypted)\n", .{});

    try stdout.print("\nDeposit Submission:\n", .{});
    try stdout.print("  [ ] Prepare deposit data with PoP\n", .{});
    try stdout.print("  [ ] Send 32 ETH to deposit contract\n", .{});
    try stdout.print("  [ ] Wait for deposit contract verification\n", .{});
    try stdout.print("  [ ] Monitor for DepositEvent emission\n", .{});
    try stdout.print("  [ ] Track validator activation on beacon chain\n", .{});

    try stdout.print("\nOngoing Operations:\n", .{});
    try stdout.print("  [ ] Use same privkey for all beacon chain signatures\n", .{});
    try stdout.print("  [ ] Never regenerate PoP (one-time at deposit)\n", .{});
    try stdout.print("  [ ] Signatures safe to aggregate with other validators\n", .{});
    try stdout.print("  [ ] Monitor for slashing conditions\n\n", .{});

    try stdout.print("=== Complete ===\n", .{});
}
